### PR TITLE
Refactor codebase and tests

### PR DESCRIPTION
**Pull request by bandantonio**
# Description

- 🗃️ Data fetching from DB: slightly improved parallelization of the fetching logic. Pretty quick, no major performance improvements.
- 🗂️ Books and highlights aggregation logic revamp - got rid of heavy array methods to reduce time complexity: O(n + m) vs O(n×m).
    - ⚡️Performance boost: ~ +50%: it is expected to have massive improvement for large datasets (for example, 100 books × 1000 annotations: 100 operations instead of 100,000).
- 🖍️ Template rendering logic - made templates compiled once and reused when passing books and annotations.
    - ⚡️Performance boost: up to +50%
- 🗄️ **New backing up logic now preserves the entire highlights folder structure including subfolders and nested files, not just direct children.**
- 🔍 Search **"From a specific book" is now almost instantaneous and filters through a list of books/authors without re-querying the database on each keystroke.**
- 🧱 Better code modularity and testability. The codebase was refactored into clear modules. Tests now benefit from fixtures, easier database mock setup without ORM dependency, and proper isolation, which made them smaller, faster and easier to maintain. **Code coverage improved significantly**.
- 🗑️ Cleanup & maintenance: Removed old db/migration scaffolding and redundant utils/method files across the repo.
- 🛠️ Tooling: Replaced [Biome](https://biomejs.dev) with [oxc tooling](https://oxc.rs) to ensure consistent linting/formatting workflow both locally and on CI, as well as improve performance.
- 🧪 Test fixes:
    - Replaced the factory-based `vi.mock` for `child_process` with `vi.mock('child_process', { spy: true })` in `dataFetching.spec.ts` for cleaner ESM-compatible mocking, avoiding the known Vitest limitation with module namespace spying.
    - Fixed a typo in a `templateProcessing.spec.ts` test name ("formtting" → "formatting").

# Checklist

- [x] I accept the [Code of Conduct](https://github.com/bandantonio/obsidian-apple-books-highlights-plugin/blob/master/.github/CODE_OF_CONDUCT.md)
- [x] I have performed a self-review of my code for consistency and potential leftover debug logs, commented out code, etc.
- [x] I have added tests that prove that my feature works or that the fix is effective
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the documentation accordingly (if necessary)
- [x] I confirm that this PR DOESN'T change the plugin's version either in the `manifest.json` or `package.json` files
- [x] I have checked that my commit messages are descriptive and follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) standard
- [x] I have enabled the checkbox to allow maintainer edits so the branch can be updated for a merge.

---

⚡ Quickly spin up Copilot coding agent tasks from anywhere on your macOS or Windows machine with [Raycast](https://gh.io/cca-raycast-docs).

### CODE REVIEW

File: src/modules/vaultManagement.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  async backupAllHighlights(): Promise<void> {
+    const highlightsFolderPath = this.getHighlightsFolderPath();
+
+    if (highlightsFolderPath) {
+      const files = (await this.vault.adapter.list(highlightsFolderPath.path)).files;
+
+      if (files.length > 0) {
+        const highlightsBackupFolderName = `${this.getHighlightsFolder()}-bk-${Date.now()}`;
+        await this.vault.adapter.rename(highlightsFolderPath.path, highlightsBackupFolderName);
+      }
```
[the remark this answers is not in the document]
Comment (bandantonio) [reply src/modules/vaultManagement.ts:61]: @copilot Good catch, but the plugin doesn't care about empty folders, so this is the expected behavior.
Following straight human logic, you want to backup something valuable in files, not just a random folder structure. So it makes sense to only check for files that were created either via full import or importing individual books. If you decide to imitate an edge case by deleting all files or recreating the highlights folder manually, backups will be skipped, as there is nothing valuable to preserve.

File: test/unit/modules/dataFetching.spec.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    };
+    const mockOn = vi.fn((event, cb) => {
+      if (event === 'close') setTimeout(() => cb(1), 0);
+      return mockOn;
+    });
+    vi.spyOn(require('child_process'), 'spawn').mockReturnValue({
+      stdout: mockStdout,
+      stderr: mockStderr,
+      on: mockOn,
+    } as any);
```
[the remark this answers is not in the document]
Comment (bandantonio) [reply test/unit/modules/dataFetching.spec.ts:215]: @copilot+claude-sonnet-4.6 Thanks. This is a [known limitation in Vitest](https://vitest.dev/guide/browser/#spying-on-module-exports).

I will add the `{ spy: true }` options as described in the documentation:

```typescript
vi.mock('child_process', { spy: true });
vi.mocked(child_process.spawn).mockReturnValue({ ... })
```
Checked locally, tests are green.

File: src/modules/dataFetching.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  } catch (error) {
+    throw new Error('No highlights found. Make sure you made some highlights in your Apple Books.', { cause: error });
+  }
+};
+
+export const dbRequest = async (dbPath: string, sqlQuery: string): Promise<IBook[]> => {
+  const dbQueryResult = spawn('sqlite3', [dbPath, sqlQuery, '-json']);
+
+  const chunks: string[] = [];
+  const errorChunks: string[] = [];
```
[the remark this answers is not in the document]
Comment (bandantonio) [reply src/modules/dataFetching.ts:91]: everything is correct here

File: test/unit/modules/dataFetching.spec.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
+import { Database } from 'better-sqlite3';
+import * as child_process from 'child_process';
+import os from 'os';
+import path from 'path';
+import { afterEach, beforeAll, beforeEach, describe, expect, test, vi } from 'vitest';
+import * as dataFetching from '../../../src/modules/dataFetching';
```
[the remark this answers is not in the document]
Comment (bandantonio) [reply test/unit/modules/dataFetching.spec.ts:6]: it's ok here, as default import throws: Cannot use namespace 'Database' as a type.

File: test/pluginDocs.spec.ts [hunk 5]
```diff
@@ -1,35 +1,53 @@
...
+
+    const allAllowedVariables = { ...allowedBookVariables, annotations: [allowedAnnotationVariables], ...allowedAnnotationVariables };
+
+    const customizationDocPath = path.join(process.cwd(), 'docs', 'customization', 'templates-and-variables.md');
+    const customizationDocContent = await fs.readFile(customizationDocPath, 'utf-8');
+
+    const templateVariablesSection = customizationDocContent.split('## Template variables')[1]?.split('## ')[0];
+    const listedVariablesInDoc = templateVariablesSection.match(/- `\{{2,3}(\w+)\}{2,3}`/gm)!;
+
+    expect(templateVariablesSection).toBeDefined();
```
[the remark this answers is not in the document]
Comment (bandantonio) [reply test/pluginDocs.spec.ts:43]: let's skip it for now

File: vitest.config.mjs [hunk 6]
```diff
@@ -11,6 +9,32 @@ export default defineConfig({
...
     ],
+    projects: [
+      {
+        test: {
+          include: ['test/unit/**/*.spec.ts'],
+          name: { label: 'unit', color: 'green' },
+          alias: {
+            obsidian: new URL('./test/mocks/obsidian.ts', import.meta.url).pathname,
+          },
+        },
```
[the remark this answers is not in the document]
Comment (bandantonio) [reply vitest.config.mjs:20]: @copilot The project is macOS-specific, no need for cross-platform compatibility
